server/pulp/server/content/storage.py

```diff
@@ -74,30 +74,45 @@ def __exit__(self, *unused):
 
 class FileStorage(ContentStorage):
     """
-    Direct storage for files and directories.
+    Direct file storage.
     """
 
-    def put(self, unit, path):
+    @staticmethod
+    def get_path(unit):
+        """
+        Get the appropriate storage path for the specified unit.
+
+        :param unit: A content unit.
+        :type unit: pulp.server.db.model.FileContentUnit
+        :return: An absolute path to where associated content files will be stored.
+        :rtype: str
+        """
+        storage_dir = os.path.join(
+            config.get('server', 'storage_dir'),
+            'content',
+            'units')
+        return os.path.join(
+            storage_dir,
+            unit.type_id,
+            unit.id[0:4],
+            unit.id)
+
+    def put(self, unit, path, location=None):
         """
         Put the content defined by the content unit into storage.
-        The file (or directory) at the specified *path* is transferred into storage.
+        The file at the specified *path* is transferred into storage.
 
         :param unit: The content unit to be stored.
         :type unit: pulp.sever.db.model.ContentUnit
         :param path: The absolute path to the file (or directory) to be stored.
         :type path: str
         """
-        storage_dir = os.path.join(
-            config.get('server', 'storage_dir'),
-            'content',
-            'units')
-        destination = os.path.join(storage_dir, unit._content_type_id, unit.id[0:4], unit.id)
+        destination = unit.storage_path
+        if location:
+            location.lstrip('/')
+            destination = os.path.join(destination, location)
         mkdir(os.path.dirname(destination))
-        if os.path.isdir(path):
-            shutil.copytree(path, destination)
-        else:
-            shutil.copy(path, destination)
-        unit._storage_path = destination
+        shutil.copy2(path, destination)
```

Comment:
@jortel use `shutil.copy` instead of `copy2`. I didn't remember which we wanted, except that we needed to be sure we used the right one. I double checked the docs just now, and what @jeremycline indicated is correct.